### PR TITLE
[Mobile] - Android - Bring the Keyboard back when closing modals

### PR DESCRIPTION
**Pull request by geriux**
This PR is part of https://github.com/WordPress/gutenberg/pull/57070

## What?
It adds new implementations in the bridge for Android, so we can control hiding and showing the keyboard in some cases.

## Why?
To improve the writing flow experience, we'll handle the keyboard manually in some cases.

## How?
It adds two methods `showAndroidSoftKeyboard` and `hideAndroidSoftKeyboard` in the React Native Bridge package, this functionality will be shared for both the demo app and the host apps as the functionality is kept entirely in `RNReactNativeGutenbergBridgeModule`.

The `showAndroidSoftKeyboard` method is introduced in the `BottomSheet` component, to prevent not displaying the Keyboard when these modals are dismissed and a TextInput was focused prior to opening the Modal.

This is only needed in Android since iOS handles this internally.

It's used in three parts of the code:

- `componentWillUnmount` some Modals get removed completely when they're hidden, which prevents other callbacks to be triggered like `onModalHide`.
- `onHardwareButtonPress` when a user dismisses the modal by using the hardware back button on Android.
- `onCloseBottomSheet` triggered when the modal is hidden either by swiping down, or tapping on the Modal backdrop.

### Case 1 - Opening the block settings of a Paragraph block
- Open the host app with metro running (It requires a local build of Gutenberg Mobile due to the bridge changes)
- Create a new post
- Add a Paragraph block
- Type some text
- Open the block settings
- Close the modal either by using the hardware back button or tapping on the backdrop
- **Expect** the Keyboard to be shown when the modal is closed

### Case 2 - Changing the text color of a Paragraph block
- Open the host app with metro running (It requires a local build of Gutenberg Mobile due to the bridge changes)
- Create a new post
- Add a Paragraph block
- Type some text
- Open the block settings
- Tap on the Text color option
- Tap on a color
- Close the modal either by using the hardware back button or tapping on the backdrop
- **Expect** the Keyboard to be shown when the modal is closed and the text color changed

### Case 3 -  Changing the text alignment
- Open the host app with metro running (It requires a local build of Gutenberg Mobile due to the bridge changes)
- Create a new post
- Add a Paragraph block
- Type some text
- Open the text alignment options
- Tap on any option
- **Expect** the Keyboard to be shown when the modal is closed and the text aligned to the previously selected option

### Case 4 - Opening the block inserter when focused on a Paragraph block
- Open the host app with metro running (It requires a local build of Gutenberg Mobile due to the bridge changes)
- Create a new post
- Add a Paragraph block
- Type some text
- Open the block insterter
- Close the modal by using the hardware back button or swiping down
- **Expect** the Keyboard to be shown when the modal is closed

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast 

Case 1|Case 2
-|-
<video src="https://github.com/WordPress/gutenberg/assets/4885740/a5682829-a993-4757-87ab-0a0b36af3410" width="200" />|<video src="https://github.com/WordPress/gutenberg/assets/4885740/212cd4b1-bb7f-4bba-8ada-253cd4c9ea24" width="200" />

Case 3|Case 4
-|-
<video src="https://github.com/WordPress/gutenberg/assets/4885740/600b2739-4f1e-422e-8e02-1cdfbedf79c7" width="200" />|<video src="https://github.com/WordPress/gutenberg/assets/4885740/4253364e-a2fa-4b88-be25-893897c9aabf" width="200" />

### CODE REVIEW

**Comment by fluiddot** [comment]
@geriux for some reason the videos captured in the PR's description don't load on my side. I wonder if there's something wrong in their source URL.

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: @geriux I noticed this delay when dismissing the bottom sheet and was wondering what would be the impact of reducing it. Seems using a value below 100 ms (when using an Android emulator), the keyboard won't be shown. Based on the inline comment, I understood that the delay was used to avoid conflicts with UI but not that it's a requirement to ensure the keyboard is displayed. What do you think about clarifying this in the comment?

On the other hand, I'm concerned that in low-end devices this value (i.e. 500ms) would be too low. I wonder if we could figure out a way to detect when the keyboard can be shown 🤔.

File: packages/react-native-bridge/index.js
Comment (fluiddot) [inline packages/react-native-bridge/index.js:524]: I saw that this function is called in cases unrelated to dismissing a bottom sheet. For example, it's called when creating/selecting Paragraph blocks, as well as selecting the post title. Specifically, the call is made from `componentWillUnmount` function of three different bottom sheets:
- `link-settings-modal`
- `block-transformations-menu`
- `block-actions-menu`
- Suggestion picker (i.e. xpost, mentions).

I haven't spotted any undesired side effects due to this, but makes me wonder if we should trigger it when bottom sheets are unmounted. I'd expect that it's only called when it's actually dismissed by a user action. Could we rely only upon the `onCloseBottomSheet` event to trigger `showAndroidSoftKeyboard` when the bottom sheet is dismissed?

**Steps:**
1. Create a post.
2. Select the first paragraph.
3. Select the post title.
4. Observe that the function is called.

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 !

I added a couple of comments but none shouldn't block the merge of the PR.

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Thanks for the feedback! I tried different approaches but it's complicated since we can easily detect when the Dialog view is fully hidden from the native side.

I made some changes in https://github.com/WordPress/gutenberg/pull/57069/commits/79d62e82b58ad6355ce1576b05a9c131ac494145

Taken from [this code](https://github.com/wordpress-mobile/WordPress-Android/blob/6e653b566c96970ebbad1f88e4574ac6a5c518fb/WordPress/src/main/java/org/wordpress/android/util/extensions/ViewExtensions.kt#L46-L79). Which worked pretty well but I had to add a delay for the main inserter menu, I think it's because is shown in Full screen and how it is implemented that from the moment it gets unmounted, the keyboard wouldn't show unless we delay it a bit.

For other and more used cases while editing (Block settings, Alignments, link settings) the keyboard is shown pretty quickly without any fixed delays.

Let me know what you think about this approach, thanks!

File: packages/react-native-bridge/index.js
Comment (geriux) [reply packages/react-native-bridge/index.js:524]: Good catch, it appears some modals are being unmounted so I added [this check](https://github.com/WordPress/gutenberg/pull/57069/commits/79d62e82b58ad6355ce1576b05a9c131ac494145#diff-7c783d6dab99fab7810467451bf92877850e8c2f4808af9dd836ca828fdd9821R224-R229).

**Comment by geriux** [comment]
> LGTM 🎊 !
> 
> I added a couple of comments but none shouldn't block the merge of the PR.

Thank you for the review and for testing these changes!

I introduced a few changes related to the keyboard logic and I also added two props for the Modal component that I think it'd improve the Android performance a bit. Let me know if you have any questions about it. Thanks!

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Nitpick: After checking [this code](https://github.com/wordpress-mobile/WordPress-Android/blob/6e653b566c96970ebbad1f88e4574ac6a5c518fb/WordPress/src/main/java/org/wordpress/android/util/extensions/ViewExtensions.kt#L46-L79), I wonder if we could extract to separate functions the logic of to simplify this function:
- `isAnyViewFocused`: Checks if there's any view focused for the current activity.
- `getCurrentFocusedView`: Returns the current focused view.

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: I have a question regarding the condition `currentActivity.getWindow().getDecorView().isShown()`, do we need to ensure that we don't show the keyboard when the editor is closed?

File: packages/react-native-bridge/index.js
Comment (fluiddot) [inline packages/react-native-bridge/index.js]: Nitpick: We could define the properties of the object:
```suggestion
 * @param {Object} options
 * @param {number} [options.delay] Delay to show soft keyboard.
```

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: > Thanks for the feedback! I tried different approaches but it's complicated since we can easily detect when the Dialog view is fully hidden from the native side.

Yeah, I imagined it wouldn't be easy. Thanks for exploring alternatives 🙇 !

> Taken from [this code](https://github.com/wordpress-mobile/WordPress-Android/blob/6e653b566c96970ebbad1f88e4574ac6a5c518fb/WordPress/src/main/java/org/wordpress/android/util/extensions/ViewExtensions.kt#L46-L79). Which worked pretty well but I had to add a delay for the main inserter menu, I think it's because is shown in Full screen and how it is implemented that from the moment it gets unmounted, the keyboard wouldn't show unless we delay it a bit.

That's interesting, I didn't know there was a similar logic in WP-Android. Regarding full-screen modals, sounds good to me to leave it as-is. In the worst case, in case the delay is not long enough for the device, we won't be showing the keyboard after dismissing the inserter modal which isn't a critical issue.

> For other and more used cases while editing (Block settings, Alignments, link settings) the keyboard is shown pretty quickly without any fixed delays.

Nice, it's great to see that there's a way to wait for the exact moment to show the keyboard 🏅.

> Let me know what you think about this approach, thanks!

I'll add some comments in the PR review to get a better understand of the logic, but the current approach looks good to me.

File: packages/react-native-bridge/index.js
Comment (fluiddot) [reply packages/react-native-bridge/index.js:524]: Ah, interesting. Thanks for addressing it.

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: I debugged the app and couldn't manage to reproduce a case where this part would be invoked. I presume we'd need it in case this function is invoked when Aztec view is focused, is this accurate?

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 ! I've tested the latest changes and worked as expected, thanks @geriux for addressing my suggestions 🙇 !

I added some minor comments but we could skip them to merge the PR. Alternatively, since the PR will be merged into a feature branch, we could address some of them directly there.

> I introduced a few changes related to the keyboard logic and I also added two props for the Modal component that I think it'd improve the Android performance a bit. Let me know if you have any questions about it. Thanks!

That's great, do you foresee we should check modals in other scenarios aside the keyboard issue due to this change?

I've tested in an actual device with a local build and can't really tell if the performance improved when showing/hiding modals. In any case, I presume that this will be more visible when testing installable builds.

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Since we allow delaying this code to be run I thought it'd be safe to check the activity is still shown. Currently, the max value we use is `500` which is enough time I guess for the editor to be closed.

I did some tests when opening the Inserter menu and then quickly tapping on the back button twice to close the editor and this code ran when it was closed as it was already scheduled.

I also saw that if we pass the value of `activeFocusedView` as `null` the `InputMethodManager` would not try to show the keyboard but it looks like it can be different depending on the Android version where it might expect a value that is not `null`.

I'd suggest keeping these checks but I would be happy to investigate further if we can remove it. What do you think?

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Nice catch! I've just tested as well and it is not being called in any of the cases when opening the different modals. I will remove that part 👍

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Good idea! I'll also extract the runnable to a separate function

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: > I did some tests when opening the Inserter menu and then quickly tapping on the back button twice to close the editor and this code ran when it was closed as it was already scheduled.

Ah, good to know that this case can be reproduced.

> I'd suggest keeping these checks but I would be happy to investigate further if we can remove it. What do you think?

My comment was merely a question to understand the code, I didn't want to imply that we should remove the code, so makes sense to me keeping the code 👍 .

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: When I was working on this I realized `mKeyboardRunnable` wasn't being set 😅 

I made a few changes in https://github.com/WordPress/gutenberg/pull/57069/commits/b0bfdf10f48440d69d585aeb379c5c9ad9fe4f56 where it splits the large function into smaller functions, with this refactor and setting the missing `mKeyboardRunnable` I realized the delay wasn't needed for the full-screen modal 🚀

Comment (fluiddot) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: > When I was working on this I realized `mKeyboardRunnable` wasn't being set 😅

Oh, I overlooked this too. Thanks for addressing it 🙇 !

> I made a few changes in https://github.com/WordPress/gutenberg/commit/b0bfdf10f48440d69d585aeb379c5c9ad9fe4f56 where it splits the large function into smaller functions, with this refactor and setting the missing mKeyboardRunnable I realized the delay wasn't needed for the full-screen modal 🚀

That's great 🎊. Was the reason for needing the delay the fact of not setting `mKeyboardRunnable` variable?

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: I noticed that in this function we have the argument `currentActivity` but in the function `getCurrentFocusedView` also used in this code it's not needed, as it's retrieved from the React context. This makes me wonder if we really need the argument. WDYT?

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: > That's great 🎊. Was the reason for needing the delay the fact of not setting mKeyboardRunnable variable?

Not necessarily it was an outcome of refactoring the whole thing 😄

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: We do need the current activity for this line https://github.com/WordPress/gutenberg/pull/57069/commits/b0bfdf10f48440d69d585aeb379c5c9ad9fe4f56#diff-3358b6e4c037a95586241526f5478f87905bd599ac2ffe341525d1a62aa1320fR590 if we remove the argument we'd need to get it from within the `createShowKeyboardRunnable` function, is this what you mean? Just to double check 😄

Comment (fluiddot) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Yeah, I thought that following the implementation of `getCurrentFocusedView` we could get the current activity from the React context. This way we wouldn't need to have the argument. Does it make sense?

https://github.com/WordPress/gutenberg/blob/b0bfdf10f48440d69d585aeb379c5c9ad9fe4f56/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java#L603

Alternatively, we could add the `currentActivity` argument to the rest of the recently incorporated functions (`getCurrentFocusedView` and `isAnyViewFocused`), so we only retrieve it in `showAndroidSoftKeyboard`.

Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Ahh I see! I've updated it in https://github.com/WordPress/gutenberg/pull/57069/commits/1db7e475989bcd180b055bd2f61c611c8d1bd91e to get the activity within `createShowKeyboardRunnable`

File: packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java
Comment (fluiddot) [inline packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: I think this condition isn't necessary because we already checked that the currently focused view is defined in line 562 via `isAnyViewFocused()`.

https://github.com/WordPress/gutenberg/blob/1db7e475989bcd180b055bd2f61c611c8d1bd91e/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java#L611-L614

Comment (geriux) [reply packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java]: Oops you're right, updated in https://github.com/WordPress/gutenberg/pull/57069/commits/1cb8d7fa13985b9c6df2e62052310e831dbc88be thanks!

**Comment by geriux** [comment]
@fluiddot Thank you for the review and feedback! I'll merge this PR once the checks pass 🚀

**Pull request merged**
